File: dopamine/packages/core/src/effects/solarbloom.ts
/**
 * Solarbloom as an `EffectFactory` on the Dopamine backbone.
 *
 * Phase 1: now DATA + a-few-lines. Its mood→params mapping + checkmark glyph
 * band live in solarbloom.dope.json (loader-resolved, byte-identical to the
 * legacy resolveParams / pickCheckGlyph). ALL renderer plumbing — program/VAO,
 * standard uniforms, the baked-SDF icon AND the font-glyph fallback texture
 * (uploaded to light + shadow), the light+shadow loop, dispose — is the shared
 * `createPassInstance` generic fullscreen-pass runner. The only code that
 * remains is the bloom SHADER + a small config naming its scalar params, its two
 * aux textures (SDF icon / glyph fallback), its shadow height, and the per-frame
 * check-draw + held-breath envelope timing.
 */

import { FRAGMENT_SRC, VERTEX_SRC } from "../engine/shader.js";
import { checkProgress, envelope } from "../engine/tempo.js";
import { MAX_MOTES, type CheckGlyph, type RenderParams } from "../engine/mood.js";
import { drawCheckGlyph } from "../engine/check-renderer.js";
import type { EffectContext, EffectFactory, EffectInstance } from "../framework/effect.js";
import { registerEffect } from "../framework/registry.js";
import { registerProgram } from "../framework/programs.js";
import { parseDope, resolveDopeParams, getOutline } from "../framework/loader.js";
import { pickBand } from "../framework/content.js";
import { createPassInstance, type AuxTextureSpec, type PassConfig, type PassParams } from "../framework/pass-runner.js";
import { decodeSdf, type DecodedSdf } from "../engine/sdf.js";
import doc from "./solarbloom.dope.json";

// Solarbloom is fully DATA-DRIVEN: its mood→params mapping lives in the bundled
// `.dope` document (solarbloom.dope.json), evaluated by the loader. A vitest
// proves the loader output is byte-identical to the legacy `resolveParams`, so
// flipping the source of truth to the file changes nothing visually.
const DOPE = parseDope(doc as object);

// GEOMETRY SEAM: the checkmark icon's SHAPE comes from the .dope's
// `geometry.outlines.checkmark.svgPath`, baked at build time into an inline SDF.
// We DECODE it once; the shader only samples it. Swapping the svgPath (+ re-bake)
// changes the rendered icon with NO shader edit. If the .dope carries no baked
// SDF we fall back to the font-glyph path, then the analytic SDF in-shader.
const CHECK_SDF: DecodedSdf | null = (() => {
  const outline = getOutline(DOPE, "checkmark");
  if (!outline?.sdf) return null;
  try {
    return decodeSdf(outline.sdf);
  } catch {
    return null;
  }
})();

// The whimsy→check-glyph fallback BANDS now live in the .dope (content.glyphBands)
// and are picked with the generic content resolver (only used when SDF is absent).
const GLYPH_BANDS = ((DOPE.content as { glyphBands?: CheckGlyph[] })?.glyphBands ?? [
  { family: "Dopamine Check Symbols", char: "✓" },
]) as CheckGlyph[];

/**
 * Resolve via the `.dope` loader → the typed RenderParams the shader consumes.
 * The numeric/palette params + the whimsy-derived CHECK GLYPH band both come
 * from the bundled `.dope` (byte-identical to `resolveParams` / `pickCheckGlyph`).
 */
function resolveFromDope(feeling: { mood: string; intensity: number; whimsy: number; seed: number }): RenderParams {
  const numeric = resolveDopeParams(DOPE, feeling, { MAX_MOTES }, "moteSeed") as unknown as RenderParams;
  return { ...numeric, checkGlyph: pickBand(GLYPH_BANDS, feeling.whimsy) };
}

// Half-size of the checkmark glyph box as a fraction of min viewport dim.
const CHECK_BOX_FRAC = 0.16;
/** Offscreen glyph texture resolution (square). Cheap; the glyph is tiny. */
const GLYPH_TEX_SIZE = 256;

const CONFIG: PassConfig = {
  vertex: VERTEX_SRC,
  fragment: FRAGMENT_SRC,
  uniforms: [
    "uCheck", "uExposure", "uBloomRadius", "uTurbulence", "uMoteSpeed",
    "uMoteCount", "uMoteSeed", "uIridescence", "uDispersion",
    "uCheckTex", "uCheckTexOn", "uCheckBox",
    "uSdfTex", "uSdfOn", "uSdfRangePx", "uSdfStrokePx",
  ],
  usesOrigin: true,
  // overshoot feeds the envelope, not a uniform; the picked glyph isn't numeric.
  bindings: { overshoot: null, moteSeed: "uMoteSeed" },
  shadowHeightFrac: (params) => (params as unknown as RenderParams).bloomRadius,
  // The checkmark box + SDF stroke px (needed by the SDF path AND the analytic).
  passUniforms: (canvas) => {
    const box = CHECK_BOX_FRAC * Math.min(canvas.width, canvas.height);
    return { uCheckBox: box, uSdfStrokePx: box * 0.11 };
  },
  auxTextures: (params) => {
    const specs: AuxTextureSpec[] = [];
    if (CHECK_SDF) {
      // GEOMETRY SEAM: the baked SDF icon takes priority (bound to TEXTURE1).
      specs.push({
        kind: "sdf",
        unit: 1,
        sdf: CHECK_SDF,
        sampler: "uSdfTex",
        onUniform: "uSdfOn",
        uniforms: (canvas) => {
          const box = CHECK_BOX_FRAC * Math.min(canvas.width, canvas.height);
          const vbW = CHECK_SDF.viewBox[2] || 100;
          return { uSdfRangePx: CHECK_SDF.range * ((2 * box) / vbW) };
        },
      });
      return specs;
    }
    // Fallback: rasterize the whimsy-chosen check glyph once into an offscreen
    // canvas and let the runner upload it as an RGBA texture (TEXTURE0). If the
    // bundled face hasn't loaded, the shader falls back to its analytic SDF.
    const g = (params as unknown as RenderParams).checkGlyph;
    if (typeof document !== "undefined") {
      const glyphCanvas = document.createElement("canvas");
      glyphCanvas.width = GLYPH_TEX_SIZE;
      glyphCanvas.height = GLYPH_TEX_SIZE;
      const gctx = glyphCanvas.getContext("2d", { alpha: true });
      if (gctx && drawCheckGlyph(gctx, GLYPH_TEX_SIZE, g.family, g.char)) {
        specs.push({ kind: "canvas", unit: 0, source: glyphCanvas, sampler: "uCheckTex", onUniform: "uCheckTexOn" });
      }
    }
    return specs;
  },
  frame: ({ animMs, life }, params) => ({
    amp: envelope(life, (params as unknown as RenderParams).overshoot),
    uCheck: checkProgress(animMs),
  }),
};

function createInstance(params: RenderParams, ctx: EffectContext): EffectInstance {
  return createPassInstance(CONFIG, params as unknown as PassParams, ctx);
}

export const solarbloom: EffectFactory<RenderParams> = {
  name: "solarbloom",
  resolve: (feeling) => resolveFromDope(feeling),
  create: createInstance,
  reducedMotion: { peakMs: 260, holdMs: 360 },
};

// Expose the renderer as a bundled PROGRAM so `loadEffect()` can bind an
// arbitrary host-authored `.dope` (one that references program "solarbloom") to
// it with no code. The numeric/palette bag comes from the loader; the whimsy-
// picked check glyph is composed on top (genuinely code-shaped, no rng).
registerProgram<RenderParams>("solarbloom", {
  create: createInstance,
  scatterKey: "moteSeed",
  consts: { MAX_MOTES },
  reducedMotion: { peakMs: 260, holdMs: 360 },
  composeParams: (numeric, feeling) => ({
    ...numeric,
    checkGlyph: pickBand(GLYPH_BANDS, feeling.whimsy),
  }),
});

export default registerEffect(solarbloom);

File: android/dopamine-effect-solarbloom/src/main/kotlin/ai/dopamine/effect/solarbloom/Solarbloom.kt
// Solarbloom as a Dopamine effect on the Android backbone — mirror of the web
// `effect-solarbloom/src/index.ts` + swift's `Solarbloom.swift`.
//
// Per the generalization mandate, the ONLY per-effect code is {the GLSL shader +
// the bespoke tempo + a tiny config naming its uniforms / bindings / shadow
// height / per-frame timing}. Everything else — the `.dope` mapping, the OKLCH
// golden-angle palette, the registry, the fullscreen-pass runner, the standard
// uniforms, the shadow geometry — is shared backbone. The numeric/palette bag
// comes verbatim from the bundled `.dope` (the SAME bytes as the web), resolved
// by the shared loader (byte-parity proven by the 192-case grid).
//
// PURE-SHADER (not a hybrid): no Canvas panel, so it uses `createPassInstance` +
// `PassConfig` (not the panel runner). The checkmark renders via the shader's
// ANALYTIC two-segment SDF branch: the GL backbone has no aux-texture support, so
// we keep uSdfOn / uCheckTexOn at 0 (the samplers uSdfTex / uCheckTex are then
// never sampled, so leaving them unbound is fine).

package ai.dopamine.effect.solarbloom

import ai.dopamine.core.DopeDoc
import ai.dopamine.core.DopeResolveInput
import ai.dopamine.core.DopeValue
import ai.dopamine.core.EffectRegistry
import ai.dopamine.core.envelope
import ai.dopamine.core.number
import ai.dopamine.core.parseDope
import ai.dopamine.core.resolveDopeParams
import ai.dopamine.gl.DrawableEffect
import ai.dopamine.gl.EffectContext
import ai.dopamine.gl.EffectInstance
import ai.dopamine.gl.PassConfig
import ai.dopamine.gl.createPassInstance
import android.content.Context
import kotlin.math.min

class Solarbloom(context: Context) : DrawableEffect {
    override val name: String = "solarbloom"

    // Load the bundled `.dope` (the EXACT web bytes) from the merged APK assets —
    // proving the data spine is shared verbatim across platforms.
    val doc: DopeDoc = parseDope(
        context.assets.open("solarbloom.dope.json").bufferedReader(Charsets.UTF_8).use { it.readText() },
    )

    override fun resolve(feeling: DopeResolveInput): Map<String, DopeValue> =
        // `MAX_MOTES` is the only const (= the GLSL `#define MAX_MOTES 80`);
        // `moteSeed` is the scatter key — both byte-identical to the web call.
        resolveDopeParams(doc, feeling, consts = mapOf("MAX_MOTES" to 80.0), scatterKey = "moteSeed")

    override fun create(params: Map<String, DopeValue>, ctx: EffectContext): EffectInstance =
        createPassInstance(CONFIG, params, ctx)

    override val reducedMotionPeakMs: Double = 260.0
    override val reducedMotionHoldMs: Double = 360.0

    companion object {
        /** Construct + register the effect (needs a Context for the `.dope` asset). */
        fun register(context: Context): Solarbloom {
            val fx = Solarbloom(context.applicationContext)
            EffectRegistry.register(fx)
            return fx
        }

        /** Half-size of the checkmark glyph box as a fraction of min viewport dim. */
        private const val CHECK_BOX_FRAC: Double = 0.16

        private val CONFIG = PassConfig(
            vertex = SOLARBLOOM_VERTEX_SRC,
            fragment = SOLARBLOOM_FRAGMENT_SRC,
            uniforms = listOf(
                "uCheck", "uExposure", "uBloomRadius", "uTurbulence", "uMoteSpeed",
                "uMoteCount", "uMoteSeed", "uIridescence", "uDispersion",
                "uCheckTex", "uCheckTexOn", "uCheckBox",
                "uSdfTex", "uSdfOn", "uSdfRangePx", "uSdfStrokePx",
            ),
            usesOrigin = true,
            // overshoot feeds the envelope, not a uniform; moteSeed drives uMoteSeed.
            bindings = mapOf("overshoot" to null, "moteSeed" to "uMoteSeed"),
            shadowHeightFrac = { p -> p.number("bloomRadius", 0.7) },
            // The checkmark box + SDF stroke px (needed by the analytic checkmark
            // path here, and the SDF path on platforms that bind it). Sized to the
            // canvas exactly like the web `passUniforms` (Math.min(w, h)). With no
            // aux-texture support we also pin uCheckTexOn / uSdfOn off so the shader
            // takes its analytic two-segment SDF branch.
            passUniforms = { widthPx, heightPx, _, _ ->
                val box = CHECK_BOX_FRAC * min(widthPx, heightPx).toDouble()
                mapOf(
                    "uCheckBox" to box.toFloat(),
                    "uSdfStrokePx" to (box * 0.11).toFloat(),
                    "uCheckTexOn" to 0f,
                    "uSdfOn" to 0f,
                )
            },
            frame = { info, params ->
                mapOf(
                    "amp" to envelope(info.life, params.number("overshoot", 1.0)).toFloat(),
                    // The check draws on its OWN ~240ms clock using the REAL elapsed
                    // time (not the "animate on twos" stepped clock) so it stays smooth
                    // even at high whimsy — matching swift's `checkProgress(info.elapsedMs)`.
                    "uCheck" to checkProgress(info.elapsedMs).toFloat(),
                )
            },
        )
    }
}
